dialer.go

```diff
@@ -538,62 +580,168 @@ func (d *Dialer) Close() error {
 	default:
 	}
 	close(d.closed)
-	d.lock.Lock()
-	defer d.lock.Unlock()
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.domainNameTicker != nil {
+		d.domainNameTicker.Stop()
+	}
 	for _, i := range d.cache {
 		i.Close()
 	}
 	return nil
 }
 
+func (d *Dialer) pollDomainNames(ctx context.Context) {
+	type cacheEntry struct {
+		cn    instance.ConnName
+		cache monitoredCache
+	}
+
+	// List all the cache entries created with a domain name
+	d.mu.RLock()
+	cacheEntries := make([]cacheEntry, 0, len(d.cache))
+	for cn, cache := range d.cache {
```

Comment:
Looks like we're copying all the keys and values out of this map into a slice. To avoid introducing a new type (cacheEntry), why don't we just allocate a new map of cn -> cache and then iterate over its keys and values? This reduces some of the complexity that adding a new type brings.